### PR TITLE
Optionally display schema in explain plan

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/11176

## Rationale for this change

In the context of:
* https://github.com/apache/datafusion/issues/9403
* https://github.com/apache/datafusion/issues/10918

I have been contemplating adding optimizations that changes DataTypes between certain operators (e.g. to DictionaryArray or StringArray)

Thus I would like to be able to see the schema in the query plans as part of that project

## What changes are included in this PR?

1. Add `set datafusion.explain.show_schema = true;` option
2. Add tests

## Are these changes tested?
Yes

## Are there any user-facing changes?

A new config option and more information in verbose explain plans

### CODE REVIEW

File: datafusion/core/src/physical_planner.rs [hunk 1]
```diff
@@ -1983,23 +1983,37 @@ impl DefaultPhysicalPlanner {
...
-                        if e.verbose && !config.show_statistics {
-                            stringified_plans.push(
-                                displayable(input.as_ref())
-                                    .set_show_statistics(true)
-                                    .to_stringified(
-                                        e.verbose,
-                                        InitialPhysicalPlanWithStats,
-                                    ),
-                            );
+                        // Show statistics + schema in verbose output even if not
```
Comment (alamb) [inline datafusion/core/src/physical_planner.rs:1994]: The amount of code required for this change is unfortunate, and I am sure we could make it better, but it was pretty mechanical

File: datafusion/sqllogictest/test_files/explain.slt [hunk 2]
```diff
@@ -259,6 +260,23 @@ physical_plan after ProjectionPushdown SAME TEXT AS ABOVE
...
+
+statement ok
+set datafusion.explain.show_schema = true;
+
+# test EXPLAIN VERBOSE
+query TT
+EXPLAIN SELECT a, b, c FROM simple_explain_test;
+----
+logical_plan TableScan: simple_explain_test projection=[a, b, c]
+physical_plan CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/data/example.csv]]}, projection=[a, b, c], has_header=true, schema=[a:Int32;N, b:Int32;N, c:Int32;N]
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/explain.slt:275]: This shows how the schema is displayed (`schema=[[a:Int32;N, b:Int32;N, c:Int32;N]]`)

**Review by jayzhan211 — APPROVED** [review]
👍

**Comment by alamb** [comment]
Thank you for the review @jayzhan211

**Pull request merged**